lib/Mojo/WebSocketProxy/Backend.pm

```diff
@@ -1,20 +1,45 @@
 package Mojo::WebSocketProxy::Backend;
-# TODO(leonerd): Split out JSON-RPC-specific parts into a subclass
 
 use strict;
 use warnings;
 
-use Mojo::Base -base;
+no indirect;
 
-use MojoX::JSON::RPC::Client;
-use Guard;
+use Mojo::Util qw(class_to_path);
 
 ## VERSION
 
-has 'url';
+our %CLASSES = (
+    jsonrpc   => 'Mojo::WebSocketProxy::Backend::JSONRPC',
+    job_async => 'Mojo::WebSocketProxy::Backend::JobAsync',
+);
+
+sub class_for_type {
+    my ($class, $type) = @_;
+    my $target_class = $CLASSES{$type};
+    # extra () around require to avoid treating the function call as a class name
+    require(class_to_path($target_class)) unless $target_class->can('new');
+    return $target_class;
+}
+
+sub new {
+    my ($class, %args) = @_;
+    return bless \%args, $class unless $class eq __PACKAGE__;
```

Comment:
Subtle. base class `->new` to find the subclass which then calls that which probably `SUPER::` dispatches back here again. Probably wants splitting into two methods here; either

    sub _really_new { return bless ... }
    sub new { do type magic }

or maybe

    sub new { return bless ... }
    sub new_for_type { my ($class, $type, %args) = @_; ... }

Would a positional first argument `$type` be appropriate, or is the caller's structure firmly dependent on `type` being a regular `%args` key?

Reply:
Normally I'd see this as a case of bad class hierarchy - a manufactured unit should not inherit the entire factory, that's just asking for a machine uprising - but a separate `->backend_instance` method might be a good option here, we wouldn't expect external users to be messing around with this class anyway.

That'd mean we have `->backend_instance($type => %args)` and a regular `->new` method, with less chance of confusion anywhere in the callchain.

Reply:
Hmm.. something like

    my $backend_instance = Mojo::WebSocketProxy::Backend->class_for_type($type)->new(%args);

maybe?